### PR TITLE
Amplify auth integration

### PR DESCRIPTION
**Pull request by suhussai**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: clients/Admin/angular.json [hunk 1]
```diff
@@ -16,74 +26,60 @@
...
-              "namedChunks": false,
-              "aot": true,
-              "extractLicenses": true,
-              "vendorChunk": false,
-              "buildOptimizer": true,
               "budgets": [
                 {
                   "type": "initial",
-                  "maximumWarning": "2mb",
+                  "maximumWarning": "1mb",
```
Comment (ujwalbukka) [inline clients/Admin/angular.json:39]: Any reason why you reduced it? I remember build failing if it was 1mb

File: clients/Admin/config-asset-loader.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { Injectable } from '@angular/core';
```
Comment (ujwalbukka) [inline clients/Admin/config-asset-loader.ts:1]: Do we need this file in reference architecture? I believe this file is for loading properties when we pre-provision build labs for workshop.

File: clients/Admin/src/app/interceptors/auth.interceptor.ts [hunk 3]
```diff
@@ -1,46 +1,46 @@
...
+  intercept(
+    req: HttpRequest<any>,
+    next: HttpHandler
+  ): Observable<HttpEvent<any>> {
+    if (req.url.includes('tenant/init')) {
+      return next.handle(req);
+    }
+    if (req.url.includes('auth-info') || req.url.includes('config.json')) {
+      return next.handle(req);
+    }
```
Comment (ujwalbukka) [inline clients/Admin/src/app/interceptors/auth.interceptor.ts:30]: Do we need this if condition?

File: clients/Admin/src/app/nav/nav.component.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  NavigationCancel,
+  NavigationEnd,
+  NavigationError,
+  NavigationStart,
+  Router,
+} from '@angular/router';
+
+import { AuthenticatorService } from '@aws-amplify/ui-angular';
+import { Auth } from 'aws-amplify';
+import { ConfigAssetLoaderService } from 'config-asset-loader';
```
Comment (ujwalbukka) [inline clients/Admin/src/app/nav/nav.component.ts:15]: Looks like we are not using ConfigAssetLoaderService

File: clients/Admin/src/aws-exports.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+const awsmobile = {
```
Comment (ujwalbukka) [inline clients/Admin/src/aws-exports.ts:1]: What is this file for?

File: clients/Admin/src/environments/environment.prod.ts [hunk 6]
```diff
@@ -1,7 +1,4 @@
 export const environment = {
   production: true,
```
Comment (ujwalbukka) [inline clients/Admin/src/environments/environment.prod.ts:2]: Do we need this file and config.json too?

File: clients/Admin/src/assets/config/config.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (ujwalbukka) [inline clients/Admin/src/assets/config/config.json:1]: For reference solution do we need this file? We will use this only for workshop right?

File: clients/Admin/yarn.lock [hunk 8]
```diff
@@ -0,0 +1,8910 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (ujwalbukka) [inline clients/Admin/yarn.lock:1]: Do we need to checkin yarn.lock file?

File: clients/Application/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,10 @@
+FROM public.ecr.aws/bitnami/node:16.16.0 AS build
```
Comment (ujwalbukka) [inline clients/Application/Dockerfile:1]: Don't think we need this file

File: clients/Application/config-asset-loader.ts [hunk 10]
```diff
@@ -0,0 +1,27 @@
+import { Injectable } from '@angular/core';
```
Comment (ujwalbukka) [inline clients/Application/config-asset-loader.ts:1]: I don't think we need this in reference solution

File: clients/Application/cypress/support/commands.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
+/// <reference types="cypress" />
```
Comment (ujwalbukka) [inline clients/Application/cypress/support/commands.ts:1]: All the contents in this file are commented out. Do we need this file?

File: clients/Application/cypress/e2e/1-getting-started/basic-access.cy.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
+/// <reference types="cypress" />
```
Comment (ujwalbukka) [inline clients/Application/cypress/e2e/1-getting-started/basic-access.cy.js:1]: Can you please add a readme under cypress folder on how to run these tests.

File: clients/Application/package.json [hunk 13]
```diff
@@ -1,73 +1,48 @@
 {
-  "name": "application",
+  "name": "dashboard",
```
Comment (ujwalbukka) [inline clients/Application/package.json:2]: Both Admin and Application client projects have name as "dashboard" is that correct?

File: clients/Application/src/app/interceptors/auth.interceptor.ts [hunk 14]
```diff
@@ -1,55 +1,46 @@
...
+  intercept(
+    req: HttpRequest<any>,
+    next: HttpHandler
+  ): Observable<HttpEvent<any>> {
+    if (req.url.includes('tenant/init')) {
+      return next.handle(req);
+    }
+    if (req.url.includes('auth-info') || req.url.includes('config.json')) {
       return next.handle(req);
+    }
```
Comment (ujwalbukka) [inline clients/Application/src/app/interceptors/auth.interceptor.ts:30]: Do we need above two conditions in Application project? I don't think so

File: clients/Application/src/app/views/error/unauthorized.component.html [hunk 15]
```diff
@@ -1,19 +1,17 @@
...
-          <div class="input-group-prepend">
-            <span class="input-group-text">
-              <i class="fa fa-envelope"></i>
-            </span>
-          </div>
-          <input
+    <form [formGroup]="tenantForm" (submit)="login()">
+      <mat-card class="card">
+        <mat-card-title>Unauthorized</mat-card-title>
+        <mat-card-subtitle>Enter your tenant name and click login below</mat-card-subtitle>
```
Comment (ujwalbukka) [inline clients/Application/src/app/views/error/unauthorized.component.html:9]: click "submit" below right?

File: clients/Application/src/app/cognito.guard.ts [hunk 16]
```diff
@@ -1,48 +1,57 @@
...
-            .join('/');
-      }
-
-      return this.oidcSecurityService.isAuthenticated$.pipe(
-          map((isAuthorized: boolean) => {
+  canActivate(
+    route: ActivatedRouteSnapshot,
+    state: RouterStateSnapshot
+  ): Promise<boolean> {
+    if (!this.authConfigService.configureAmplifyAuth()) {
```
Comment (ujwalbukka) [inline clients/Application/src/app/cognito.guard.ts:26]: We need to check whether user is authenticated or not. Maybe checking for Auth.currentSession() might be appropriate here?

File: clients/Application/src/app/views/auth/auth-configuration.service.ts [hunk 17]
```diff
@@ -14,167 +14,82 @@
...
-            autoUserinfo: false,
-
-          });
-
-        }
-
+  configureAmplifyAuth(): boolean {
+    try {
+      const userPoolId = localStorage.getItem('userPoolId');
+      const appClientId = localStorage.getItem('appClientId');
```
Comment (ujwalbukka) [inline clients/Application/src/app/views/auth/auth-configuration.service.ts:72]: If these values are not present, does it through an exception?

File: clients/Application/src/app/views/error/unauthorized.component.ts [hunk 18]
```diff
@@ -55,31 +63,35 @@ export class UnauthorizedComponent implements OnInit {
...
-
-    });
+      this.openErrorMessageSnackBar(this.errorMessage);
+      return false;
+    }
+
+    this.authConfigService
+      .setTenantConfig(tenantName)
+      .then((val) => {
+        this.router.navigate(['/dashboard']);
```
Comment (ujwalbukka) [inline clients/Application/src/app/views/error/unauthorized.component.ts:87]: How is it routing to login? by routing to '/dashboard'

File: clients/Application/src/app/views/orders/detail/detail.component.html [hunk 19]
```diff
@@ -110,7 +121,7 @@ <h3>Services / Products</h3>
       </div>
     </div>
     <div class="invoice-footer">
-      Thank you for choosing the Serverless SaaS Reference Architecture.
+      Thank you for choosing the SaaS Serverless Reference Architecture.
```
Comment (ujwalbukka) [inline clients/Application/src/app/views/orders/detail/detail.component.html:124]: should be "Serverless SaaS"

File: clients/Application/src/assets/config/config.json [hunk 20]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (ujwalbukka) [inline clients/Application/src/assets/config/config.json:1]: I don't think we need this file for reference solution.

File: clients/Application/yarn.lock [hunk 21]
```diff
@@ -0,0 +1,9545 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (ujwalbukka) [inline clients/Application/yarn.lock:1]: Need this file?

File: clients/Landing/yarn.lock [hunk 22]
```diff
@@ -0,0 +1,9524 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (ujwalbukka) [inline clients/Landing/yarn.lock:1]: need this file?

File: deployment.sh [hunk 23]
```diff
@@ -1,219 +1,217 @@
...
-npm install --legacy-peer-deps && npm run build
+cat <<EoF >./src/aws-exports.ts
+const awsmobile = {
+    "aws_project_region": "$REGION",
+    "aws_cognito_region": "$REGION",
+    "aws_user_pools_id": "$ADMIN_USERPOOLID",
+    "aws_user_pools_web_client_id": "$ADMIN_APPCLIENTID",
+};
+
+export default awsmobile;
```
Comment (ujwalbukka) [inline deployment.sh:112]: ??

File: server/bootstrap-template.yaml [hunk 24]
```diff
@@ -218,6 +218,11 @@ Outputs:
     Value: !GetAtt Cognito.Outputs.CognitoOperationUsersUserPoolClientId
     Export:
       Name: "Serverless-SaaS-AdminUserPoolClientId" 
+  CognitoOperationUsersUserPoolId:
```
Comment (ujwalbukka) [inline server/bootstrap-template.yaml:221]: I guess we can remove "Serverless-SaaS-AdminUserPoolProviderURL" output parameter

File: clients/Application/cypress/e2e/1-getting-started/basic-access.cy.js (hunk 12 above)
Comment (suhussai) [reply clients/Application/cypress/e2e/1-getting-started/basic-access.cy.js:1]: Add info on running in headless and headed mode.

File: clients/Application/src/app/interceptors/auth.interceptor.ts (hunk 14 above)
Comment (suhussai) [reply clients/Application/src/app/interceptors/auth.interceptor.ts:30]: Remove these lines;
```
if (req.url.includes('auth-info') || req.url.includes('config.json')) {
      return next.handle(req);
    }
```

File: clients/Application/src/app/views/auth/auth-configuration.service.ts (hunk 17 above)
Comment (suhussai) [reply clients/Application/src/app/views/auth/auth-configuration.service.ts:72]: Add a comment to ensure that `Amplify.configure()` will fail if not all values are present.

Comment (suhussai) [reply clients/Application/src/app/views/auth/auth-configuration.service.ts:72]: Added a check in lines 74 and 75 to ensure that we never run `Amplify.configure()` with empty values:
```
      if (!userPoolId || !appClientId) {
        return false;
      }
```

File: clients/Application/cypress/e2e/1-getting-started/basic-access.cy.js (hunk 12 above)
Comment (suhussai) [reply clients/Application/cypress/e2e/1-getting-started/basic-access.cy.js:1]: added a README under the cypress folder

File: clients/Application/src/app/cognito.guard.ts (hunk 16 above)
Comment (suhussai) [reply clients/Application/src/app/cognito.guard.ts:26]: discussed offline. We need to first ensure that Amplify has been configured before checking the session.

File: clients/Application/src/app/views/error/unauthorized.component.ts (hunk 18 above)
Comment (suhussai) [reply clients/Application/src/app/views/error/unauthorized.component.ts:87]: `<amplify-authenticator>` components takes care of checking for a session and routing the user to the login page when needed.

**Pull request merged**